# k-Means Clustering

In [1]:
# Standard imports
import numpy as np
import matplotlib.pyplot as plt

# Generate fake data
from sklearn.datasets import load_wine

# For model and evaluation
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score, adjusted_rand_score
from sklearn.metrics import classification_report, accuracy_score
from sklearn.model_selection import train_test_split

# Suppress KMeans warnings
from warnings import filterwarnings

In [2]:
# Load the Wine dataset
data = load_wine()
X = data.data
y = data.target

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=21)

In [3]:
clusters = 3

In [4]:
# Suppress warnings
filterwarnings("ignore", category=FutureWarning, module="sklearn")

# Create a KMeans clustering model
kmeans = KMeans(n_clusters=clusters, random_state=22)
kmeans.fit(X_train)

# Predict cluster labels
labels = kmeans.labels_

# Classification of testing set
y_pred = kmeans.predict(X_test)

In [5]:
# Evaluate clustering using silhouette score
silhouette_avg = silhouette_score(X_train, labels)
print(f'Silhouette Score: {silhouette_avg:.2f}')

# Evaluate clustering using adjusted Rand index
rand_index = adjusted_rand_score(y_test, y_pred)
print(f'Adjusted Rand Index: {rand_index:.2f}')

# Evaluate classification using the classification report
classification_rep = classification_report(y_test, y_pred)
print('Classification Report:\n', classification_rep)

Silhouette Score: 0.56
Adjusted Rand Index: 0.33
Classification Report:
               precision    recall  f1-score   support

           0       0.33      0.27      0.30        15
           1       0.00      0.00      0.00        10
           2       0.46      0.55      0.50        11

    accuracy                           0.28        36
   macro avg       0.26      0.27      0.27        36
weighted avg       0.28      0.28      0.28        36



# Write Up 

Performing k-means clustering on the wine dataset in Scikit-Learn, an unsupervised learning technique, yielded interesting results. The objective of this approach was to group the data into three clusters, reflecting potential underlying structures within the dataset.

The metrics obtained from this k-means clustering analysis are as follows:

**Silhouette Score: 0.56**
The silhouette score measures the quality of clustering, with higher values indicating better-defined clusters. A score of 0.56 suggests that the data points within the clusters are reasonably close to each other and well-separated from other clusters, indicating a moderate level of clustering quality.

**Adjusted Rand Index: 0.33**
The Adjusted Rand Index (ARI) quantifies the similarity between the true class labels (in this case, three wine classes) and the cluster assignments. An ARI of 0.33 suggests that the clustering results do show some degree of agreement with the ground truth labels, but the agreement is moderate.

However, when we evaluate the clustering results using a classification report, it becomes apparent that the k-means clustering approach faces some challenges when compared to supervised classification models:

**Classification Report:**

               precision    recall  f1-score   support

           0       0.33      0.27      0.30        15
           1       0.00      0.00      0.00        10
           2       0.46      0.55      0.50        11

    accuracy                           0.28        36
    macro avg      0.26      0.27      0.27        36
    weighted avg   0.28      0.28      0.28        36

Here, precision, recall, and F1-score values are reported for each cluster. The results show that the clustering approach struggles to achieve high precision and recall, especially for cluster 1, which has precision and recall values of 0.00. This indicates that cluster 1 does not align well with any of the true wine classes.

The overall accuracy of 0.28 highlights that the clustering results do not strongly match the true class labels, indicating that k-means clustering, in this instance, may not be the most suitable approach for this dataset when compared to supervised classification models.

In summary, while the silhouette score and adjusted Rand index suggest moderate clustering quality and some degree of agreement with true labels, the classification report reveals challenges in accurately clustering the data into three classes. Generally, when we have access to the actual labels, supervised learning methods will perform significantly better than unsupervised methods. 